Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -5384,6 +4290,241 @@ - (void)presentViewController:(UIViewController *)viewController
     [super presentViewController:viewController animated:animated completion:completion];
 }
 
+#pragma mark - ConversationViewModelDelegate
+
+- (void)conversationViewModelWillUpdate
+{
+    OWSAssertIsOnMainThread();
+    OWSAssertDebug(self.conversationViewModel);
+
+    OWSLogVerbose(@"");
+
+    // HACK to work around radar #28167779
+    // "UICollectionView performBatchUpdates can trigger a crash if the collection view is flagged for layout"
+    // more: https://github.com/PSPDFKit-labs/radar.apple.com/tree/master/28167779%20-%20CollectionViewBatchingIssue
+    // This was our #2 crash, and much exacerbated by the refactoring somewhere between 2.6.2.0-2.6.3.8
+    //
+    // NOTE: It's critical we do this before beginLongLivedReadTransaction.
+    //       We want to relayout our contents using the old message mappings and
+    //       view items before they are updated.
+    [self.collectionView layoutIfNeeded];
+    // ENDHACK to work around radar #28167779
+}
+
+- (void)conversationViewModelDidUpdate:(ConversationUpdate *)conversationUpdate
```

Comment:
Corresponds to the old "db did update" handler.  Note that we can receive 3 kinds of updates:

* Minor (no changes to conversation items, but may need to update other subviews like the nav bar).
* Diff (an incremental update; some items were added/removed/updated).
* Reload (incremental update isn't possible, reload the entire view).